app/gui2/src/composables/events.ts

```diff
@@ -288,71 +288,81 @@ export const enum PointerButtonMask {
+    trackedElement?.releasePointerCapture(trackedPointer)
 
-      lastPos = null
-      trackedElement = null
+    if (handler(computePosition(e, initialGrabPos, lastPos), e, 'stop') !== false) {
+      e.stopImmediatePropagation()
     }
-    trackedPointer.value = null
+
+    current.value = undefined
   }
 
   function doMove(e: PointerEvent) {
-    if (trackedElement != null && initialGrabPos != null && lastPos != null) {
-      if (handler(computePosition(e, initialGrabPos, lastPos), e, 'move') !== false) {
-        e.stopImmediatePropagation()
-      }
-      lastPos = new Vec2(e.clientX, e.clientY)
+    if (current.value?.trackedPointer !== e.pointerId) return
+    const { initialGrabPos, lastPos } = current.value
+
+    if (handler(computePosition(e, initialGrabPos, lastPos), e, 'move') !== false) {
+      e.stopImmediatePropagation()
     }
+    current.value.lastPos = new Vec2(e.clientX, e.clientY)
   }
 
   const events = {
     pointerdown(e: PointerEvent) {
       // pointers should not respond to unmasked mouse buttons
-      if ((e.buttons & requiredButtonMask) === 0 || (predicate && !predicate(e))) {
-        return
-      }
+      if ((e.buttons & requiredButtonMask) === 0 || !predicate(e)) return
+      if (current.value != null) return
 
-      if (trackedPointer.value == null && e.currentTarget instanceof Element) {
-        trackedPointer.value = e.pointerId
-        // This is mostly SAFE, as virtually all `Element`s also extend `GlobalEventHandlers`.
-        trackedElement = e.currentTarget as Element & GlobalEventHandlers
+      let trackedElement: Element | undefined
+      const trackedTarget =
+        pointerCapturedBy === 'currentTarget' ? e.currentTarget
+        : pointerCapturedBy === 'target' ? e.target
+        : null
+      if (trackedTarget instanceof Element) {
         // `setPointerCapture` is not defined in tests.
-        trackedElement.setPointerCapture?.(e.pointerId)
-        initialGrabPos = new Vec2(e.clientX, e.clientY)
-        lastPos = initialGrabPos
-        if (handler(computePosition(e, initialGrabPos, lastPos), e, 'start') !== false) {
-          e.stopImmediatePropagation()
-        }
+        trackedTarget?.setPointerCapture?.(e.pointerId)
+        trackedElement = trackedTarget
+      }
+      const initialGrabPos = new Vec2(e.clientX, e.clientY)
+      const lastPos = initialGrabPos
+      current.value = {
+        trackedPointer: e.pointerId,
+        initialGrabPos,
+        lastPos,
+        trackedElement,
+      }
+      if (handler(computePosition(e, initialGrabPos, lastPos), e, 'start') !== false) {
+        e.stopImmediatePropagation()
       }
     },
     pointerup(e: PointerEvent) {
-      if (trackedPointer.value !== e.pointerId) {
-        return
-      }
       doStop(e)
     },
     pointermove(e: PointerEvent) {
-      if (trackedPointer.value !== e.pointerId) {
-        return
-      }
```

Comment:
Removing those early reaturns will cause issues when you have multiple pointers (e.g. multitouch screen fingers or a pen). You don't want to update the drag when unrelated pointer is lifted or moved.

Reply:
They're duplicated: both `doStop` and `doMove` check `trackedPointer`.